### PR TITLE
FIX: skip category preference update if already set by group.

### PR DESCRIPTION
**Pull request by vinothkannans**
`default_categories_*` site settings will update the category preferences on user create. But it should update user's category preference if a group's setting already updated it for that user.

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -2046,6 +2046,10 @@ def hash(password, salt)
     fab!(:category3) { Fabricate(:category) }
     fab!(:category4) { Fabricate(:category) }
 
+    # UGLY but perf is horrible with this callback
+    before { User.set_callback(:create, :after, :ensure_in_trust_level_group) }
+    after { User.skip_callback(:create, :after, :ensure_in_trust_level_group) }
```
Comment (tgxworld) [inline spec/models/user_spec.rb:2051]: Is this related to the change here? If not, I think splitting out into another PR where the before/after timings are provided is more useful.

File: app/models/user.rb [hunk 2]
```diff
@@ -1968,7 +1968,9 @@ def set_default_categories_preferences
       end
     end
 
-    CategoryUser.insert_all!(values) if values.present?
+    if values.present?
+      CategoryUser.insert_all(values, unique_by: :idx_category_users_user_id_category_id)
```
Comment (tgxworld) [inline app/models/user.rb:1972]: Referencing an index of another table in the `User` model is quite strange because I think the details about the indexes of a table should be considered an internal concern. Are we able to do an `UPSERT ... ON CONCLICT DO NOTHING` here instead?

File: spec/models/user_spec.rb (hunk 1 above)
Comment (vinothkannans) [reply spec/models/user_spec.rb:2051]: Yes, else the tests will fail since the `ensure_in_trust_level_group` callback is skipped by default.

File: app/models/user.rb (hunk 2 above)
Comment (vinothkannans) [reply app/models/user.rb:1972]: Or we can do `unique_by: %i[user_id category_id]`. Isn't `ON CONFLICT DO NOTHING` will ignore all the exceptions?

Comment (tgxworld) [reply app/models/user.rb:1972]: Yup that sounds much better :+1:

File: spec/models/user_spec.rb (hunk 1 above)
Comment (tgxworld) [reply spec/models/user_spec.rb:2051]: Can we perhaps run `Group.user_trust_level_change!(id, trust_level)` for the user manually and only for the test which you've added?

Comment (vinothkannans) [reply spec/models/user_spec.rb:2051]: It's impossible since I need to run this before the `after_create` callback `set_default_categories_preferences`. Instead, I can create the record directly using the `CategoryUser.create` method.

File: app/models/user.rb [hunk 3]
```diff
@@ -1968,7 +1968,7 @@ def set_default_categories_preferences
       end
     end
 
-    CategoryUser.insert_all!(values) if values.present?
+    CategoryUser.insert_all(values) if values.present?
```
Comment (vinothkannans) [inline app/models/user.rb:1971]: I find out that if you didn't specify the `unique_key:` param then it will use the default unique key available in the table.

**Review by tgxworld — APPROVED** [review]

**Comment by tgxworld** [comment]
Thank you for the changes :+1: It looks good
